File: src/dp/longest_inc_subsequence.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int lengthOfLIS(vector<int> nums) {
	int n = (int)nums.size();
	vector<int> res(n, 1);
	for (int i = n - 1; i >= 0; i--) {
		for (int j = i + 1; j < n; j++) {
			if (nums[j] > nums[i]) {
				res[i] = max(res[i], res[j] + 1);
			}
		}
	}
	return *max_element(res.begin(), res.end());
}

int main() {
	cout << lengthOfLIS({1, 2, 4, 3}) << endl; // 3
}

File: src/dp/longest_inc_subsequence.py
def lengthOfLIS(nums):
    dp = [1] * len(nums)
    for i in range(len(nums)-1, -1, -1):
        for j in range(i+1, len(nums)):
            if (nums[j] > nums[i]):
                dp[i] = max(dp[i], 1 + dp[j])
    return max(dp)

print(lengthOfLIS([1,2,4,3])) # 3

"""
longest increacing subsequence
the longest increacing subsequence at
any given index is 1 because itself
would be a part of the increasing subsequence
worst case scenario there's no way to
"""
